File: algorithms.py
"""
Core graph-search algorithms for AlgoPath: BFS, DFS, and Dijkstra.

Each function returns a `SearchResult` containing the path found, its total
cost, and an `operations` counter (number of nodes popped/visited) so the
CLI can report a rough empirical comparison of algorithm cost alongside
their theoretical Big-O complexity.
"""

from collections import deque
from dataclasses import dataclass, field

from .graph import Graph
from .priority_queue import PriorityQueue


@dataclass
class SearchResult:
    algorithm: str
    path: list[str]
    cost: float
    operations: int  # number of nodes expanded/visited
    complexity: str
    found: bool = field(init=False)

    def __post_init__(self):
        self.found = len(self.path) > 0


def _reconstruct_path(came_from: dict, start: str, goal: str) -> list[str]:
    if goal not in came_from and goal != start:
        return []
    path = [goal]
    while path[-1] != start:
        prev = came_from.get(path[-1])
        if prev is None:
            return []
        path.append(prev)
    path.reverse()
    return path


def bfs(graph: Graph, start: str, goal: str) -> SearchResult:
    """Breadth-First Search: shortest path by number of edges (unweighted)."""
    visited = {start}
    came_from: dict[str, str] = {}
    queue = deque([start])
    operations = 0

    while queue:
        node = queue.popleft()
        operations += 1
        if node == goal:
            break
        for neighbor, _weight in graph.neighbors(node):
            if neighbor not in visited:
                visited.add(neighbor)
                came_from[neighbor] = node
                queue.append(neighbor)

    path = _reconstruct_path(came_from, start, goal)
    cost = sum(graph.adjacency[a][b] for a, b in zip(path, path[1:])) if path else 0
    return SearchResult("BFS", path, cost, operations, "O(V + E)")


def dfs(graph: Graph, start: str, goal: str) -> SearchResult:
    """Depth-First Search: finds *a* path (not guaranteed shortest)."""
    visited = set()
    came_from: dict[str, str] = {}
    stack = [start]
    operations = 0

    while stack:
        node = stack.pop()
        if node in visited:
            continue
        visited.add(node)
        operations += 1
        if node == goal:
            break
        for neighbor, _weight in graph.neighbors(node):
            if neighbor not in visited:
                came_from.setdefault(neighbor, node)
                stack.append(neighbor)

    path = _reconstruct_path(came_from, start, goal)
    cost = sum(graph.adjacency[a][b] for a, b in zip(path, path[1:])) if path else 0
    return SearchResult("DFS", path, cost, operations, "O(V + E)")


def dijkstra(graph: Graph, start: str, goal: str) -> SearchResult:
    """Dijkstra's algorithm: guaranteed shortest weighted path (non-negative weights)."""
    distances = {node: float("inf") for node in graph.nodes()}
    distances[start] = 0
    came_from: dict[str, str] = {}
    visited = set()
    operations = 0

    pq = PriorityQueue()
    pq.push(0, start)

    while not pq.is_empty():
        dist, node = pq.pop_min()
        if node in visited:
            continue
        visited.add(node)
        operations += 1

        if node == goal:
            break

        for neighbor, weight in graph.neighbors(node):
            new_dist = dist + weight
            if new_dist < distances.get(neighbor, float("inf")):
                distances[neighbor] = new_dist
                came_from[neighbor] = node
                pq.push(new_dist, neighbor)

    path = _reconstruct_path(came_from, start, goal)
    cost = distances.get(goal, float("inf"))
    if cost == float("inf"):
        cost = 0
    return SearchResult("Dijkstra", path, cost, operations, "O((V + E) log V)")


ALGORITHMS = {
    "bfs": bfs,
    "dfs": dfs,
    "dijkstra": dijkstra,
}

File: Algorithms.java
import java.util.*;

public class Algorithms {

    public static class SearchResult {
        public final String algorithm;
        public final List<String> path;
        public final double cost;
        public final int operations;
        public final String complexity;
        public final boolean found;

        public SearchResult(String algorithm, List<String> path, double cost,
                             int operations, String complexity) {
            this.algorithm = algorithm;
            this.path = path;
            this.cost = cost;
            this.operations = operations;
            this.complexity = complexity;
            this.found = !path.isEmpty();
        }

        @Override
        public String toString() {
            String pathStr = found ? String.join(" -> ", path) : "(no path)";
            return String.format("%-9s found=%-6s cost=%-8s expanded=%-5d complexity=%-16s path=%s",
                    algorithm, found, cost, operations, complexity, pathStr);
        }
    }

    private static List<String> reconstructPath(Map<String, String> cameFrom, String start, String goal) {
        if (!cameFrom.containsKey(goal) && !goal.equals(start)) return new ArrayList<>();
        LinkedList<String> path = new LinkedList<>();
        String current = goal;
        path.addFirst(current);
        while (!current.equals(start)) {
            String prev = cameFrom.get(current);
            if (prev == null) return new ArrayList<>();
            path.addFirst(prev);
            current = prev;
        }
        return path;
    }

    private static double pathCost(Graph graph, List<String> path) {
        double cost = 0;
        for (int i = 0; i < path.size() - 1; i++) {
            cost += graph.neighbors(path.get(i)).get(path.get(i + 1));
        }
        return cost;
    }

    /** Breadth-First Search — shortest path by edge count (unweighted). */
    public static SearchResult bfs(Graph graph, String start, String goal) {
        Set<String> visited = new HashSet<>(List.of(start));
        Map<String, String> cameFrom = new HashMap<>();
        Deque<String> queue = new ArrayDeque<>();
        queue.add(start);
        int operations = 0;

        while (!queue.isEmpty()) {
            String node = queue.poll();
            operations++;
            if (node.equals(goal)) break;
            for (String neighbor : graph.neighbors(node).keySet()) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    cameFrom.put(neighbor, node);
                    queue.add(neighbor);
                }
            }
        }
        List<String> path = reconstructPath(cameFrom, start, goal);
        double cost = path.isEmpty() ? 0 : pathCost(graph, path);
        return new SearchResult("BFS", path, cost, operations, "O(V + E)");
    }

    /** Depth-First Search — finds *a* path, not guaranteed shortest. */
    public static SearchResult dfs(Graph graph, String start, String goal) {
        Set<String> visited = new HashSet<>();
        Map<String, String> cameFrom = new HashMap<>();
        Deque<String> stack = new ArrayDeque<>();
        stack.push(start);
        int operations = 0;

        while (!stack.isEmpty()) {
            String node = stack.pop();
            if (visited.contains(node)) continue;
            visited.add(node);
            operations++;
            if (node.equals(goal)) break;
            for (String neighbor : graph.neighbors(node).keySet()) {
                if (!visited.contains(neighbor)) {
                    cameFrom.putIfAbsent(neighbor, node);
                    stack.push(neighbor);
                }
            }
        }
        List<String> path = reconstructPath(cameFrom, start, goal);
        double cost = path.isEmpty() ? 0 : pathCost(graph, path);
        return new SearchResult("DFS", path, cost, operations, "O(V + E)");
    }

    /** Dijkstra's algorithm using Java's built-in PriorityQueue (binary heap). */
    public static SearchResult dijkstra(Graph graph, String start, String goal) {
        Map<String, Double> dist = new HashMap<>();
        for (String n : graph.nodes()) dist.put(n, Double.POSITIVE_INFINITY);
        dist.put(start, 0.0);
        Map<String, String> cameFrom = new HashMap<>();
        Set<String> visited = new HashSet<>();
        int operations = 0;

        PriorityQueue<double[]> pq = new PriorityQueue<>(Comparator.comparingDouble(a -> a[0]));
        Map<Integer, String> idToNode = new HashMap<>();
        Map<String, Integer> nodeToId = new HashMap<>();
        int idCounter = 0;
        for (String n : graph.nodes()) {
            idToNode.put(idCounter, n);
            nodeToId.put(n, idCounter);
            idCounter++;
        }
        pq.add(new double[]{0.0, nodeToId.get(start)});

        while (!pq.isEmpty()) {
            double[] top = pq.poll();
            String node = idToNode.get((int) top[1]);
            if (visited.contains(node)) continue;
            visited.add(node);
            operations++;
            if (node.equals(goal)) break;

            for (var entry : graph.neighbors(node).entrySet()) {
                String neighbor = entry.getKey();
                double weight = entry.getValue();
                double newDist = top[0] + weight;
                if (newDist < dist.getOrDefault(neighbor, Double.POSITIVE_INFINITY)) {
                    dist.put(neighbor, newDist);
                    cameFrom.put(neighbor, node);
                    pq.add(new double[]{newDist, nodeToId.get(neighbor)});
                }
            }
        }

        List<String> path = reconstructPath(cameFrom, start, goal);
        double cost = dist.getOrDefault(goal, Double.POSITIVE_INFINITY);
        if (Double.isInfinite(cost)) cost = 0;
        return new SearchResult("Dijkstra", path, cost, operations, "O((V + E) log V)");
    }
}
